README.md
![CI](https://github.com/shlima/health_bit/workflows/CI/badge.svg)
[![gem version](https://badge.fury.io/rb/health_bit.svg)](https://rubygems.org/gems/health_bit)

# HealthBit

![](./doc/logo.png?sanitize=true)

This gem was inspired by the [health_check](https://github.com/ianheggie/health_check), but is simpler and more
extensible and contains up to 95% less code.

Key differences:
* is a rack application (just a lambda function)
* can be used with rails, sinatra or any other rack application
* can add custom checks
* can add multiple endpoints with independent checks
* can use any rack middleware (such as http basic auth, IP whitelist)

## Toc

* [Installation](#installation)
* [Configuration](#configuration)
* [Add Checks](#add-checks)
* [Add a Route](#add-a-route)
* [Password Protection](#password-protection)
* [Multiple endpoints](#multiple-endpoints)
* [Custom formatter](#custom-formatter)

## Check Examples

* [Database check](#database-check)
* [Redis check](#redis-check)
* [Sidekiq check](#sidekiq-check)
* [Rails cache check](#rails-cache-check)
* [Elasticsearch check](#elasticsearch-check)
* [RabbitMQ check](#rabbitmq-check)
* [HTTP check](#http-check)
* [ClickHouse check](#clickhouse-check)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'health_bit'
```

## Configuration

```ruby
# config/initializers/health_bit.rb

HealthBit.configure do |c|
  # DEFAULT SETTINGS ARE SHOWN BELOW
  c.success_text = '%<count>d checks passed 🎉'
  c.headers = {
    'Content-Type' => 'text/plain;charset=utf-8',
    'Cache-Control' => 'private,max-age=0,must-revalidate,no-store'
  }
  c.success_code = 200
  c.fail_code = 500
  c.show_backtrace = false
  c.formatter = HealthBit::Formatter.new
  # DEFAULT SETTINGS ARE SHOWN ABOVE

  c.add('Check name') do
    # Body check, should returns `true`
    true
  end
end
```

## Add Checks

By default, the **gem does not contain any checks**, **you should add the
necessary checks by yourself**. The check should return `false` or `nil`
to be considered unsuccessful or throw an exception, any other
values are considered satisfactory.

Example checks:

```ruby
## Successful checks
HealthBit.add('PostgreSQL') do |env|
  ApplicationRecord.connection.select_value('SELECT 1') == 1
end

HealthBit.add('Custom') do |env|
  next(false) if 1 != 0 
  
  true
end

## Failed checks
HealthBit.add('Database') do |env|
  false
end

HealthBit.add('Docker service') do |env|
  raise 'not responding'
end

# The Check can be added as an object responding to a call
# (to be able to test your check)
class Covid19Check
  def self.call(env)
    false
  end
end

HealthBit.add('COVID-19 Checker', Covid19Check)
```

## Add a Route

Since the gem is a rack application, you must mount it to app's
routes. Below is an example for the Rails.

```ruby
# config/routes.rb

Rails.application.routes.draw do
  mount HealthBit.rack => '/health'
end
```

```bash
curl --verbose http://localhost:3000/health

< HTTP/1.1 200 OK
< Content-Type: text/plain;charset=utf-8
< Cache-Control: private,max-age=0,must-revalidate,no-store
< X-Request-Id: 59a796b9-29f7-4302-b1ff-5d0b06dd6637
< X-Runtime: 0.006007
< Vary: Origin
< Transfer-Encoding: chunked

4 checks passed 🎉
```

## Password Protection

Since the gem is a common rack application, you can add any rack
middleware to it. Below is an example with HTTP-auth for the Rails.

```ruby
# config/routes.rb

Rails.application.routes.draw do
  HealthBit.rack.use Rack::Auth::Basic do |username, password|
    ActiveSupport::SecurityUtils.secure_compare(Digest::SHA256.hexdigest(username), Digest::SHA256.hexdigest('user')) & ActiveSupport::SecurityUtils.secure_compare(Digest::SHA256.hexdigest(password), Digest::SHA256.hexdigest('password'))
  end

  mount HealthBit.rack => '/health'
end
```

## Database check

```ruby
HealthBit.add('Database') do |env|
  ApplicationRecord.connection.select_value('SELECT 1') == 1
end
```

## Redis check

```ruby
HealthBit.add('Redis') do |env|
  Redis.current.ping == 'PONG'
end
```

## Sidekiq check

```ruby
HealthBit.add('Sidekiq') do |env|
  Sidekiq.redis(&:ping) == 'PONG'
end
```

## Rails cache check

```ruby
HealthBit.add('Rails cache') do |env|
  Rails.cache.write('__health_bit__', '1', expires_in: 1.second)
end
```

## Elasticsearch check

```ruby
HealthBit.add('Elasticsearch') do |env|
  Elasticsearch::Client.new.ping
end
```

## RabbitMQ check

```ruby
HealthBit.add('RabbitMQ') do |env|
  Bunny::Connection.connect(&:connection)
end
```

## HTTP check

```ruby
HealthBit.add('HTTP check') do |env|
  Net::HTTP.new('www.example.com', 80).request_get('/').kind_of?(Net::HTTPSuccess)
end
```

## ClickHouse check

```ruby
HealthBit.add('ClickHouse') do |env|
  ClickHouse.connection.ping
end
```

## Multiple endpoints

Sometimes you have to add several health check endpoints. Let's say
you have to check the docker container health and the health
of your application as a whole. Below is an example for the Rails.

```ruby
# config/initializers/health_bit.rb

DockerCheck = HealthBit.clone
AppCheck = HealthBit.clone

DockerCheck.add('Docker Health') do |env|
  true
end

AppCheck.add('App Health') do |env|
  ApplicationRecord.connection.select_value("SELECT 't'::boolean")
end
```

```ruby
# config/routes.rb

Rails.application.routes.draw do
  mount DockerCheck.rack => '/docker'
  mount AppCheck.rack => '/app'
end
```

## Custom formatter

You can easily [configure](https://github.com/shlima/health_bit/blob/master/lib/health_bit/formatter.rb) custom format of response body, headers and
http statuses.

```ruby
class JsonFormatter < HealthBit::Formatter
  # @param error HealthBit::CheckError
  # @param env Hash
  # @param health_bit HealthBit
  def format_failure(error, env, health_bit)
    { 'status': 'error' }.to_json
  end

  # @param error HealthBit::CheckError
  # @param env Hash
  # @param health_bit HealthBit
  def headers_failure(error, env, health_bit)
    {
      'Content-Type' => 'application/json'
    }
  end
end

HealthBit.configure do |c|
  c.formatter = JsonFormatter.new
end
```


CHANGELOG.md
## 0.1.9
- [13](https://github.com/shlima/health_bit/pull/13) Pass env to check blocks or #call methods
- [11](https://github.com/shlima/health_bit/issues/11) Added custom formatter


spec/spec_helper.rb
require 'bundler/setup'
require 'health_bit'
require 'pry'

RSpec.configure do |config|
  config.example_status_persistence_file_path = '.rspec_status'
  config.disable_monkey_patching!
  config.filter_run :focus
  config.run_all_when_everything_filtered = true
  config.order = :random
  Kernel.srand config.seed

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/health_bit/check_spec.rb
RSpec.describe HealthBit::Check do
  describe '#call' do
    context 'when fail' do
      let(:subjects) do
        [
          described_class.new('foo', -> { nil }),
          described_class.new('foo', -> { false }),
          described_class.new('foo', -> { raise }),
          described_class.new('foo', -> { break(false); true }),
          described_class.new('foo', Class.new { def self.call ; false ; end }),
          described_class.new('foo', Class.new { def self.call ; nil ; end }),
          described_class.new('foo', Class.new { def self.call ; raise ; end }),
        ]
      end

      it 'returns an error' do
        subjects.each do |subject|
          expect(subject.call).to be_a(HealthBit::CheckError)
        end
      end
    end

    context 'when success' do
      let(:subjects) do
        [
          described_class.new('bar', -> { true }),
          described_class.new('bar', -> { break(true); false }),
          described_class.new('bar', Class.new { def self.call ; true ; end }),
        ]
      end

      it 'returns an error' do
        subjects.each do |subject|
          expect(subject.call).to eq(nil)
        end
      end
    end

    context 'env' do
      let(:subjects) do
        [
          described_class.new('foo', -> (env) { env.call }),
          described_class.new('foo', Class.new { def self.call(env) ; env.call ; end }),
          described_class.new('foo', Proc.new { |env| env.call  }),
        ]
      end


      it 'passes env to a handler' do
        subjects.each do |subject|
          env = double('student')
          expect(env).to receive(:call).once
          subject.call(env)
        end
      end
    end
  end
end


spec/health_bit/health_bit_spec.rb
RSpec.describe HealthBit do
  subject do
    described_class.clone
  end

  describe '#configure' do
    let(:configure) do
      this = nil
      subject.configure { |o| this = o }
      this
    end

    it 'yields self' do
      expect(configure).to eq(subject)
    end
  end

  describe '#success_code, #success_code=' do
    context 'when default' do
      it 'works' do
        expect(subject.success_code).to eq(200)
      end
    end

    context 'when set' do
      before do
        subject.success_code = 0
      end

      it 'works' do
        expect(subject.success_code).to eq(0)
      end
    end
  end

  describe '#fail_code, #fail_code=' do
    context 'when default' do
      it 'works' do
        expect(subject.fail_code).to eq(500)
      end
    end

    context 'when set' do
      before do
        subject.fail_code = 0
      end

      it 'works' do
        expect(subject.fail_code).to eq(0)
      end
    end
  end

  describe '#formatter, #formatter=' do
    context 'when default' do
      it 'works' do
        expect(subject.formatter).to be_a(HealthBit::Formatter)
      end
    end

    context 'when set' do
      before do
        subject.formatter = 0
      end

      it 'works' do
        expect(subject.formatter).to eq(0)
      end
    end
  end

  describe '#headers, #headers=' do
    context 'when default' do
      it 'works' do
        expect(subject.headers).to eq({'Content-Type' => 'text/plain;charset=utf-8', 'Cache-Control' => 'private,max-age=0,must-revalidate,no-store' })
      end
    end

    context 'when modified' do
      it 'returns a clone (for the Grape)' do
        expect { subject.headers.clear }.not_to change { subject.headers }
      end
    end

    context 'when set' do
      before do
        subject.headers = { foo: :bar }
      end

      it 'works' do
        expect(subject.headers).to eq(foo: :bar)
      end
    end
  end

  describe '#success_text, #success_text=' do
    context 'when default' do
      it 'works' do
        expect(subject.success_text).to eq('0 checks passed 🎉')
      end
    end

    context 'when set' do
      before do
        subject.success_text = 'foo'
      end

      it 'works' do
        expect(subject.success_text).to eq('foo')
      end
    end
  end

  describe '#add' do
    it 'returns self' do
      expect(subject.add('1', Class.new)).to eq(subject)
    end

    context 'when object' do
      before do
        subject.add('foo', Class.new)
      end

      it 'adds check' do
        expect(subject.checks.length).to eq(1)
        expect(subject.checks.first).to have_attributes(name: 'foo')
      end
    end

    context 'when block' do
      before do
        subject.add('bar') do
          1
        end
      end

      it 'works' do
        expect(subject.checks.length).to eq(1)
        expect(subject.checks.first).to have_attributes(name: 'bar')
      end
    end

    context 'when both handler and block passed' do
      let(:add) do
        subject.add('bar', Class.new) do
          1
        end
      end

      it 'errors' do
        expect { add }.to raise_error(ArgumentError)
      end
    end

    context 'when neither handler nor block passed' do
      let(:add) do
        subject.add('bar')
      end

      it 'errors' do
        expect { add }.to raise_error(ArgumentError)
      end
    end
  end

  describe '#rack' do
    let(:response) do
      subject.rack.call({}).to_s
    end

    context 'when show_backtrace=true' do
      before do
        subject.show_backtrace = true
        subject.add('foo') { false }
      end

      it 'prints backtrace' do
        expect(response).to include('health_bit/lib/health_bit')
      end
    end

    context 'when show_backtrace=false' do
      before do
        subject.add('foo') { false }
      end

      it 'hides backtrace' do
        expect(response).not_to include('health_bit/lib/health_bit')
      end
    end

    context 'when adding middleware' do
      it 'works' do
        expect { subject.rack.use(Class.new) }.to change { subject.rack.instance_variable_get(:@use).count }.by(1)
      end
    end
  end

  describe '#clone' do
    let!(:donor) do
      subject.clone
    end

    let!(:dolly_1) do
      donor.clone
    end

    let!(:dolly_2) do
      dolly_1.clone
    end

    context 'when #checks' do
      before do
        donor.add('donor', 1)
        dolly_1.add('dolly', 1)
      end

      it 'works' do
        expect(donor.checks.length).to eq(1)
        expect(dolly_1.checks.length).to eq(1)
        expect(dolly_2.checks).to be_empty
        expect(donor.checks.first).to have_attributes(name: 'donor')
        expect(dolly_1.checks.first).to have_attributes(name: 'dolly')
      end
    end

    context 'when attr_accessors (like success_text)' do
      before do
        donor.success_text = 'donor succeed'
        dolly_1.success_text = 'dolly succeed'
      end

      it 'works' do
        expect(dolly_2.success_text).to eq(described_class.success_text)
        expect(dolly_1.success_text).to eq('dolly succeed')
        expect(donor.success_text).to eq('donor succeed')
      end
    end
  end
end


lib/health_bit/formatter.rb
# frozen_string_literal: true

require 'rack'
require 'health_bit/version'

module HealthBit
  # rubocop:disable Lint/UnusedMethodArgument
  class Formatter
    # @param error HealthBit::CheckError
    # @param env Hash
    # @param health_bit HealthBit
    def format_success(error, env, health_bit)
      health_bit.success_text
    end

    # @param error HealthBit::CheckError
    # @param env Hash
    # @param health_bit HealthBit
    def format_failure(error, env, health_bit)
      format = health_bit.show_backtrace ? CheckError::FORMAT_FULL : CheckError::FORMAT_SHORT
      error.to_s(format)
    end

    # @param env Hash
    # @param health_bit HealthBit
    def headers_success(env, health_bit)
      health_bit.headers
    end

    # @param error HealthBit::CheckError
    # @param env Hash
    # @param health_bit HealthBit
    def headers_failure(error, env, health_bit)
      health_bit.headers
    end

    # @param env Hash
    # @param health_bit HealthBit
    def code_success(env, health_bit)
      health_bit.success_code
    end

    # @param error HealthBit::CheckError
    # @param env Hash
    # @param health_bit HealthBit
    def code_failure(error, env, health_bit)
      health_bit.fail_code
    end
  end
  # rubocop:enable Lint/UnusedMethodArgument
end


lib/health_bit.rb
# frozen_string_literal: true

require 'rack'
require 'health_bit/version'

module HealthBit
  autoload :Check, 'health_bit/check'
  autoload :CheckError, 'health_bit/check_error'
  autoload :Formatter, 'health_bit/formatter'

  DEFAULT_SUCCESS_TEXT = '%<count>d checks passed 🎉'
  DEFAULT_HEADERS = {
    'Content-Type' => 'text/plain;charset=utf-8',
    'Cache-Control' => 'private,max-age=0,must-revalidate,no-store'
  }.freeze
  DEFAULT_SUCCESS_CODE = 200
  DEFAULT_FAIL_CODE = 500
  DEFAULT_FORMATTER = Formatter.new

  extend self # rubocop:disable Style/ModuleFunction

  attr_writer :success_text, :success_code, :fail_code, :headers, :formatter
  attr_accessor :show_backtrace

  def success_text
    format(@success_text || DEFAULT_SUCCESS_TEXT, count: checks.length)
  end

  def success_code
    @success_code || DEFAULT_SUCCESS_CODE
  end

  def fail_code
    @fail_code || DEFAULT_FAIL_CODE
  end

  def headers
    (@headers || DEFAULT_HEADERS).dup
  end

  # @return [Formatter]
  def formatter
    @formatter || DEFAULT_FORMATTER
  end

  def checks
    @checks ||= []
  end

  def configure
    yield(self)
  end

  # @return [self]
  def add(name, handler = nil, &block)
    raise ArgumentError, <<~MSG if handler && block
      Both <handler> and <block> were passed to the <#{name}> check
    MSG

    raise ArgumentError, <<~MSG unless handler || block
      Nor <handler> or <block> were passed to the <#{name}> check
    MSG

    checks.push(Check.new(name, handler || block))

    self
  end

  # @return [nil, CheckError]
  def check(env)
    checks.each do |check|
      (exception = check.call(env)).nil? ? next : (return exception)
    end

    nil
  end

  def rack(this = self)
    @rack ||= Rack::Builder.new do
      run ->(env) do
        if (error = this.check(env))
          [
            this.formatter.code_failure(error, env, this),
            this.formatter.headers_failure(error, env, this),
            [this.formatter.format_failure(error, env, this)]
          ]
        else
          [
            this.formatter.code_success(env, this),
            this.formatter.headers_success(env, this),
            [this.formatter.format_success(error, env, this)]
          ]
        end
      end
    end
  end

  def clone
    Module.new.tap do |dolly|
      dolly.singleton_class.include(HealthBit)
    end
  end
end
